### PR TITLE
Adding another way of calculating gradients in downsamplers

### PR DESCRIPTION
**Pull request by XianzheMa**
This PR addresses https://github.com/eth-easl/modyn/issues/505.
A flag is introduced, to switch between using purely the last layer gradients, or the concatenation of last layer gradients and penultimate layer gradients, in selection algorithms.

### CODE REVIEW

File: modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_matrix_downsampling_strategy.py [hunk 1]
```diff
@@ -9,7 +9,11 @@
 )
 from modyn.trainer_server.internal.trainer.remote_downsamplers.deepcore_utils.shuffling import _shuffle_list_and_tensor
 
-MatrixContent = Enum("MatrixContent", ["EMBEDDINGS", "GRADIENTS"])
+
+class MatrixContent(Enum):
+    EMBEDDINGS = 1
+    LAST_LAYER_GRADIENTS = 2
+    LAST_TWO_LAYERS_GRADIENTS = 3
```
Comment (XianzheMa) [inline modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_matrix_downsampling_strategy.py:17]: The previous way is not the canonical way to write an enum

File: modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_per_label_remote_downsample_strategy.py [hunk 2]
```diff
@@ -6,17 +6,6 @@
...
-        self,
-        pipeline_id: int,
-        trigger_id: int,
-        batch_size: int,
-        params_from_selector: dict,
-        modyn_config: dict,
-        device: str,
-    ):
-        super().__init__(pipeline_id, trigger_id, batch_size, params_from_selector, modyn_config, device)
-        self.requires_data_label_by_label = True
```
Comment (XianzheMa) [inline modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_per_label_remote_downsample_strategy.py:19]: many subclasses still overwrite this field `requires_data_label_by_label` to `False`. Thus I leave it for subclass to decide. (actually only Craig is the only subclass that previous does not overwrite this field itself.)

File: modyn/trainer_server/internal/trainer/remote_downsamplers/remote_craig_downsampling.py [hunk 3]
```diff
@@ -43,13 +46,16 @@ def __init__(
...
+
         if self.greedy not in OPTIMIZER_CHOICES:
             raise ValueError(
                 f"The required Greedy optimizer is not available. Pick one of the following: {OPTIMIZER_CHOICES}"
             )
 
-        # This class uses the embedding recorder
-        self.requires_coreset_supporting_module = True
+        self.requires_coreset_supporting_module = self.full_grad_approximation == "LastLayerWithEmbedding"
+        self.requires_data_label_by_label = True
```
Comment (XianzheMa) [inline modyn/trainer_server/internal/trainer/remote_downsamplers/remote_craig_downsampling.py:58]: Craig always selects points label by label. Previously this field is set implicitly to True in the base class `AbstractPerLabelRemoteDownsamplingStrategy`

File: analytics/tools/aggregate_runs/core_aggregation.py [hunk 4]
```diff
@@ -3,7 +3,6 @@
 from pathlib import Path
 
 import pandas as pd
-
```
Comment (MaxiBoether) [inline analytics/tools/aggregate_runs/core_aggregation.py:6]: nit, but no need to touch this file in this PR :D

File: modyn/config/schema/pipeline/sampling/downsampling_config.py [hunk 5]
```diff
@@ -74,6 +77,7 @@ class GradMatchDownsamplingConfig(BaseDownsamplingConfig):
 
     strategy: Literal["GradMatch"] = "GradMatch"
     balance: bool = Field(False, description="If True, the samples are balanced.")
+    full_grad_approximation: FullGradApproximation = Field(default="LastLayer")
```
Comment (MaxiBoether) [inline modyn/config/schema/pipeline/sampling/downsampling_config.py:80]: Could you either here or at the definition on `FullGradApproximation` add a description/comment on the meaning of this such that future users understand what this is doing?

File: modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_craig_remote_downsampling.py [hunk 6]
```diff
@@ -20,6 +20,7 @@ def get_sampler_config(modyn_config, balance=False):
         "balance": balance,
         "selection_batch": 64,
         "greedy": "NaiveGreedy",
+        "full_grad_approximation": "LastLayerWithEmbedding",
```
Comment (MaxiBoether) [inline modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_craig_remote_downsampling.py:23]: This affects all downsamplers: Do we have a test somewhere at least that we do not crash when using `LastLayer` instead? we don't need to write a test where we manually calculate something and then assert equivalence but having some form of validation it does not crash would be nice

File: modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_matrix_downsampling_strategy.py [hunk 7]
```diff
@@ -61,12 +68,26 @@ def inform_samples(
...
         assert self.matrix_content is not None
-
-        if self.matrix_content == MatrixContent.GRADIENTS:
-            new_elements = self._compute_gradients(forward_output, target, embedding)
+        if self.matrix_content == MatrixContent.LAST_LAYER_GRADIENTS:
+            grads_wrt_loss_sum = (
+                self._compute_last_layer_gradient_wrt_loss_sum(self.criterion, forward_output, target).detach().cpu()
+            )
+            grads_wrt_loss_mean = grads_wrt_loss_sum / batch_size
+            new_elements = grads_wrt_loss_mean.detach().cpu()
```
Comment (MaxiBoether) [inline modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_matrix_downsampling_strategy.py:78]: no need to detach again, I think

File: modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_per_label_remote_downsample_strategy.py (hunk 2 above)
Comment (MaxiBoether) [reply modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_per_label_remote_downsample_strategy.py:19]: sounds good!

File: modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_remote_downsampling_strategy.py [hunk 8]
```diff
@@ -101,3 +103,51 @@ def select_points(self) -> tuple[list[int], torch.Tensor]:
...
+        embedding_dim = embedding.shape[1]
+        num_classes = forward_output.shape[1]
+        batch_num = target.shape[0]
+
+        with torch.no_grad():
+            bias_parameters_grads = torch.autograd.grad(loss, forward_output)[0]
+            weight_parameters_grads = embedding.view(batch_num, 1, embedding_dim).repeat(
+                1, num_classes, 1
+            ) * bias_parameters_grads.view(batch_num, num_classes, 1).repeat(1, 1, embedding_dim)
+            gradients = torch.cat([bias_parameters_grads, weight_parameters_grads.flatten(1)], dim=1)
```
Comment (MaxiBoether) [inline modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_remote_downsampling_strategy.py:152]: I guess I am still confused where this every calculates the gradient for the second-to-last layer , I see that we have d loss / d output but then I only see .view and .repeat operations where I don't get how this would calculate a gradient, but this is also not really within this PR

**Review by MaxiBoether — APPROVED** [review]
Thanks, my only concern is (and I think codecov complains as well) that we should at least test whether the new code path does not crash or something. But the changes itself look good and I tried to validate that the old logic stay the same :)

File: modyn/config/schema/pipeline/sampling/downsampling_config.py (hunk 5 above)
Comment (XianzheMa) [reply modyn/config/schema/pipeline/sampling/downsampling_config.py:80]: Make sense. Done

File: modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_matrix_downsampling_strategy.py (hunk 7 above)
Comment (XianzheMa) [reply modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_matrix_downsampling_strategy.py:78]: thanks for spotting it

File: modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_craig_remote_downsampling.py (hunk 6 above)
Comment (XianzheMa) [reply modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_craig_remote_downsampling.py:23]: make sense. will do

**Comment by MaxiBoether** [comment]
Thanks, the new tests look good. Feel free to merge!
